File: packages/core/src/descriptor.ts
import type {
  SlopNode, Affordance, JsonSchema, ActionHandler,
  NodeDescriptor, ItemDescriptor, Action, ParamDef,
} from "./types";

export interface NormalizationResult {
  node: SlopNode;
  handlers: Map<string, ActionHandler>;
}

/**
 * Convert a developer-friendly NodeDescriptor into a wire-format SlopNode
 * and extract action handlers into a flat map keyed by "{path}/{action}".
 */
export function normalizeDescriptor(
  path: string,
  id: string,
  descriptor: NodeDescriptor
): NormalizationResult {
  const handlers = new Map<string, ActionHandler>();
  const children: SlopNode[] = [];

  // Convert items → children with type "item"
  if (descriptor.items) {
    for (const item of descriptor.items) {
      const itemPath = path ? `${path}/${item.id}` : item.id;
      const { node: itemNode, handlers: itemHandlers } = normalizeItem(itemPath, item);
      children.push(itemNode);
      for (const [k, v] of itemHandlers) handlers.set(k, v);
    }
  }

  // Convert inline children → children (recursive)
  if (descriptor.children) {
    for (const [childId, childDesc] of Object.entries(descriptor.children)) {
      const childPath = path ? `${path}/${childId}` : childId;
      const { node: childNode, handlers: childHandlers } = normalizeDescriptor(
        childPath, childId, childDesc
      );
      children.push(childNode);
      for (const [k, v] of childHandlers) handlers.set(k, v);
    }
  }

  // Convert actions → affordances + extract handlers
  const affordances = normalizeActions(path, descriptor.actions, handlers);

  const node: SlopNode = {
    id,
    type: descriptor.type,
    ...(descriptor.props && { properties: descriptor.props }),
    ...(children.length > 0 && { children }),
    ...(affordances.length > 0 && { affordances }),
    ...(descriptor.meta && { meta: descriptor.meta }),
  };

  return { node, handlers };
}

function normalizeItem(
  path: string,
  item: ItemDescriptor
): NormalizationResult {
  const handlers = new Map<string, ActionHandler>();
  const children: SlopNode[] = [];

  // Item can have inline children too
  if (item.children) {
    for (const [childId, childDesc] of Object.entries(item.children)) {
      const childPath = `${path}/${childId}`;
      const { node, handlers: h } = normalizeDescriptor(childPath, childId, childDesc);
      children.push(node);
      for (const [k, v] of h) handlers.set(k, v);
    }
  }

  const affordances = normalizeActions(path, item.actions, handlers);

  const node: SlopNode = {
    id: item.id,
    type: "item",
    ...(item.props && { properties: item.props }),
    ...(children.length > 0 && { children }),
    ...(affordances.length > 0 && { affordances }),
    ...(item.meta && { meta: item.meta }),
  };

  return { node, handlers };
}

function normalizeActions(
  path: string,
  actions: Record<string, Action> | undefined,
  handlers: Map<string, ActionHandler>
): Affordance[] {
  if (!actions) return [];
  const affordances: Affordance[] = [];

  for (const [name, action] of Object.entries(actions)) {
    const handlerKey = path ? `${path}/${name}` : name;

    if (typeof action === "function") {
      // Bare callback
      handlers.set(handlerKey, action);
      affordances.push({ action: name });
    } else {
      // Full ActionDescriptor
      handlers.set(handlerKey, action.handler);
      affordances.push({
        action: name,
        ...(action.label && { label: action.label }),
        ...(action.description && { description: action.description }),
        ...(action.dangerous && { dangerous: true }),
        ...(action.idempotent && { idempotent: true }),
        ...(action.params && { params: normalizeParams(action.params) }),
      });
    }
  }

  return affordances;
}

/** Convert simplified params { title: "string" } to JSON Schema */
function normalizeParams(params: Record<string, ParamDef>): JsonSchema {
  const properties: Record<string, JsonSchema> = {};
  const required: string[] = [];

  for (const [key, def] of Object.entries(params)) {
    if (typeof def === "string") {
      properties[key] = { type: def };
    } else {
      properties[key] = {
        type: def.type,
        ...(def.description && { description: def.description }),
        ...(def.enum && { enum: def.enum }),
      };
    }
    required.push(key);
  }

  return { type: "object", properties, required };
}

File: packages/python/slop-ai/src/slop/descriptor.py
"""Normalize developer-friendly dict descriptors into wire-format SlopNodes.

Developer descriptors use plain dicts with keys like ``props``, ``actions``,
``items``, ``children``.  This module converts them into proper ``SlopNode``
instances and extracts a flat ``{path/action: handler}`` map.
"""

from __future__ import annotations

from typing import Any, Callable

from .types import Affordance, ContentRef, NodeMeta, SlopNode


ActionHandler = Callable[..., Any]


def normalize_descriptor(
    path: str,
    node_id: str,
    descriptor: dict[str, Any],
) -> tuple[SlopNode, dict[str, ActionHandler]]:
    """Convert a descriptor dict into a ``SlopNode`` and handler map."""
    handlers: dict[str, ActionHandler] = {}
    children: list[SlopNode] = []

    # Build meta
    meta_dict: dict[str, Any] = dict(descriptor.get("meta") or {})
    if "summary" in descriptor:
        meta_dict["summary"] = descriptor["summary"]

    # Windowed collection
    window_desc = descriptor.get("window")
    if window_desc is not None:
        for item in window_desc["items"]:
            item_path = f"{path}/{item['id']}" if path else item["id"]
            item_node, item_handlers = _normalize_item(item_path, item)
            children.append(item_node)
            handlers.update(item_handlers)
        meta_dict["total_children"] = window_desc["total"]
        meta_dict["window"] = (window_desc["offset"], len(window_desc["items"]))
    elif "items" in descriptor and descriptor["items"] is not None:
        for item in descriptor["items"]:
            item_path = f"{path}/{item['id']}" if path else item["id"]
            item_node, item_handlers = _normalize_item(item_path, item)
            children.append(item_node)
            handlers.update(item_handlers)

    # Inline children (recursive)
    inline_children = descriptor.get("children")
    if inline_children:
        for child_id, child_desc in inline_children.items():
            child_path = f"{path}/{child_id}" if path else child_id
            child_node, child_handlers = normalize_descriptor(child_path, child_id, child_desc)
            children.append(child_node)
            handlers.update(child_handlers)

    # Actions → affordances + handlers
    affordances = _normalize_actions(path, descriptor.get("actions"), handlers)

    # Properties
    props = descriptor.get("props")
    properties = dict(props) if props else None

    # Content ref as top-level field (per spec 13)
    cr: ContentRef | None = None
    cr_raw = descriptor.get("content_ref") or descriptor.get("contentRef")
    if cr_raw:
        ref = dict(cr_raw)
        if "uri" not in ref:
            ref["uri"] = f"slop://content/{path}"
        cr = ContentRef.from_dict(ref)

    meta = NodeMeta.from_dict(meta_dict) if meta_dict else None

    node = SlopNode(
        id=node_id,
        type=descriptor["type"],
        properties=properties or None,
        children=children or None,
        affordances=affordances or None,
        meta=meta,
        content_ref=cr,
    )
    return node, handlers


def _normalize_item(
    path: str,
    item: dict[str, Any],
) -> tuple[SlopNode, dict[str, ActionHandler]]:
    """Convert an item descriptor dict into a SlopNode."""
    handlers: dict[str, ActionHandler] = {}
    children: list[SlopNode] = []

    # Item inline children
    inline_children = item.get("children")
    if inline_children:
        for child_id, child_desc in inline_children.items():
            child_path = f"{path}/{child_id}"
            child_node, child_handlers = normalize_descriptor(child_path, child_id, child_desc)
            children.append(child_node)
            handlers.update(child_handlers)

    affordances = _normalize_actions(path, item.get("actions"), handlers)

    meta_dict: dict[str, Any] = dict(item.get("meta") or {})
    if "summary" in item:
        meta_dict["summary"] = item["summary"]
    meta = NodeMeta.from_dict(meta_dict) if meta_dict else None

    node = SlopNode(
        id=item["id"],
        type="item",
        properties=item.get("props") or None,
        children=children or None,
        affordances=affordances or None,
        meta=meta,
    )
    return node, handlers


def _normalize_actions(
    path: str,
    actions: dict[str, Any] | None,
    handlers: dict[str, ActionHandler],
) -> list[Affordance]:
    """Convert an actions dict into a list of Affordances, populating handlers."""
    if not actions:
        return []

    affordances: list[Affordance] = []
    for name, action in actions.items():
        handler_key = f"{path}/{name}" if path else name

        if callable(action):
            handlers[handler_key] = action
            affordances.append(Affordance(action=name))
        else:
            # Dict-style action descriptor
            handlers[handler_key] = action["handler"]
            affordances.append(Affordance(
                action=name,
                label=action.get("label"),
                description=action.get("description"),
                dangerous=action.get("dangerous", False),
                idempotent=action.get("idempotent", False),
                estimate=action.get("estimate"),
                params=_normalize_params(action["params"]) if "params" in action else None,
            ))

    return affordances


def _normalize_params(params: dict[str, Any]) -> dict[str, Any]:
    """Convert simplified params ``{"title": "string"}`` into JSON Schema."""
    properties: dict[str, Any] = {}
    required: list[str] = []

    for key, defn in params.items():
        if isinstance(defn, str):
            properties[key] = {"type": defn}
        else:
            prop: dict[str, Any] = {"type": defn["type"]}
            if "description" in defn:
                prop["description"] = defn["description"]
            if "enum" in defn:
                prop["enum"] = defn["enum"]
            properties[key] = prop
        required.append(key)

    return {"type": "object", "properties": properties, "required": required}
